src/modules/stack_overflow.py

```diff
@@ -0,0 +1,80 @@
+import discord.ext.commands as commands 
+import discord.embeds as embed
+import discord.colour as colour
+import requests, stackexchange 
+from logger import ErrorLevel
+from bs4 import BeautifulSoup
+
+''' ISSUES: 
+1) beautiful soup removes all links (<a href>) in answer_body
+2) include question body as well?  
+'''
+class StackOverflow(commands.Cog):
+    def __init__(self, bot, cfg, logger):
+        self.bot = bot
+        self.logger = logger
+        mod = cfg.modules.stack_overflow
+        self.engine_id = mod.googlesearch_engine_id
+        self.key = mod.googlesearch_key
+
+    # uses Discord's ```  ``` code block markdown
+    def markUp(self, str):
+        return str.replace("<pre><code>","```").replace("</code></pre>","```")
+            
+    ''' Uses google search API first because Stack Overflow API can only return questions that exactly match the search terms'''
+    def findAnswer(self, search_term):
+        try:
+            search_term = '%20'.join(search_term.split(" "))
+
+            r = requests.get("https://www.googleapis.com/customsearch/v1/siterestrict?q=" + search_term + "&cx=" + self.engine_id + "&key=" + self.key)
+            json_content = r.json()
+            question_url = json_content["items"][0]['link']
+            id = int(question_url.split("/")[4])
+            so = stackexchange.Site(stackexchange.StackOverflow)
+            question = so.question(id, body=True)
+            question_title = "Q: " + question.title
+            question_body = question.body
+            #question_body = self.markUp(BeautifulSoup(question.body).get_txt())
+            answers = question.answers
+            answer_body = ""
+            for answer in answers:
+                if (answer.accepted == True):
+                    answer_body = answer.body
+
+            # no accepted answer, grab highest voted one instead 
+            if (answer_body == ""):
+                answers.sort(key = lambda answer: answer.score, reverse=True)
+                answer_body = answers[0].body
+
+            answer_body = self.markUp(answer_body)
+            soup = BeautifulSoup(markup=answer_body,features="html.parser")
+            answer_body = soup.get_text()
+            answer_body = "Top answer: " + answer_body 
+            if (len(answer_body) > 2048):
+                answer_body = answer_body[:2045] + "..."
+                # max length is 2048
+
+
+            return question_title, question_body, answer_body, question_url
+
+        except NameError:
+            return -1
+
+        except:
+            return -2
+
```

Comment:
this function does way too much in my opinion, id break it down if you can, also the returning of -1 and -2 is kinda unclear, are these status codes for the terminal?